src/hotspot/share/nmt/vmtCommon.cpp

```diff
@@ -0,0 +1,151 @@
+    address committed_top = crgn.base() + crgn.size();
+    if (committed_top < top) {
+      // committed stack guard pages, skip them
+      bottom = crgn.base() + crgn.size();
+    } else {
+      assert(top == committed_top, "Sanity, top=" INTPTR_FORMAT " , com-top=" INTPTR_FORMAT, p2i(top), p2i(committed_top));
+      return false;;
+    }
+    return true;
+  });
+
+  return bottom;
+}
+
+// Iterate the range, find committed region within its bound.
+class RegionIterator : public StackObj {
+private:
+  const address _start;
+  const size_t  _size;
+
+  address _current_start;
+public:
+  RegionIterator(address start, size_t size) :
+    _start(start), _size(size), _current_start(start) {
+  }
+
+  // return true if committed region is found
+  bool next_committed(address& start, size_t& size);
+private:
+  address end() const { return _start + _size; }
+};
+
+bool RegionIterator::next_committed(address& committed_start, size_t& committed_size) {
+  if (end() <= _current_start) return false;
+
+  const size_t page_sz = os::vm_page_size();
+  const size_t current_size = end() - _current_start;
+  if (os::committed_in_range(_current_start, current_size, committed_start, committed_size)) {
+    assert(committed_start != nullptr, "Must be");
+    assert(committed_size > 0 && is_aligned(committed_size, os::vm_page_size()), "Must be");
+
+    _current_start = committed_start + committed_size;
+    return true;
+  } else {
+    return false;
+  }
+}
+
+// Walk all known thread stacks, snapshot their committed ranges.
+class SnapshotThreadStackWalker : public VirtualMemoryWalker {
+public:
+  SnapshotThreadStackWalker() {}
+
+  bool do_allocation_site(const ReservedMemoryRegion* rgn) {
+    if (rgn->mem_tag() == mtThreadStack) {
+      address stack_bottom = rgn->thread_stack_uncommitted_bottom();
+      address committed_start;
+      size_t  committed_size;
+      size_t stack_size = rgn->base() + rgn->size() - stack_bottom;
+      // Align the size to work with full pages (Alpine and AIX stack top is not page aligned)
+      size_t aligned_stack_size = align_up(stack_size, os::vm_page_size());
+
+      ReservedMemoryRegion* region = const_cast<ReservedMemoryRegion*>(rgn);
+      NativeCallStack ncs; // empty stack
+
+      RegionIterator itr(stack_bottom, aligned_stack_size);
+      DEBUG_ONLY(bool found_stack = false;)
+      while (itr.next_committed(committed_start, committed_size)) {
+        assert(committed_start != nullptr, "Should not be null");
+        assert(committed_size > 0, "Should not be 0");
+        // unaligned stack_size case: correct the region to fit the actual stack_size
+        if (stack_bottom + stack_size < committed_start + committed_size) {
+          committed_size = stack_bottom + stack_size - committed_start;
+        }
+        VirtualMemoryTracker::Instance::add_committed_region(committed_start, committed_size, ncs);
```

Comment:
Is it a problem that it is never recorded that regions are uncommitted?  Committed regions are added on-the-fly when reporting data, and those regions persist after reporting is done. So maybe it's possible to report too much committed size on a future report. The old implementation had this characteristic too.

Reply:
Uncommit for mtThreadStack are tracked when a thread exits and cleans/releases its stack.
Here, the committed pages that are not reported yet to NMT are extracted.